### DISCUSSION TITLE
Getting rid of ThreadLocals

### DISCUSSION THREAD

**Discussion by robertbraeutigam** [open-telemetry/opentelemetry-java#6584] [Q&A]
At this point the SDK is riddled with usages of `ThreadLocal`s. Sometimes this is easier to spot, when using `current()` or similar methods. Sometimes it is implicit and very hard to spot.

Using `ThreadLocal`s makes it very difficult and error-prone to use this SDK in reactive designs, when using `CompletableFuture`s, or when using it  in Scala with "cats-effect" or similar effect library.

I propose that that core library should not use `ThreadLocal` at all. At most it should be a (clearly marked) optional layer on top.

**Comment by jkwatson** [comment]
I think "riddled with" is overstating it quite a bit [I see precisely 9 instantiations of ThreadLocals in the entire codebase (outside of tests)]. Almost all of the ThreadLocal usage are local caches, and hence not critical to functionality, just performance. 

You are free to use a different ContextStorage implementation that doesn't use ThreadLocal  if you want to. That's really the only functionally-relevant ThreadLocal that I know of.

**Reply by jkwatson** [reply]
In addition, I have never had any problems using our API and SDK with CompletableFuture. You just need to make sure that the ExecutorService that is used is one that is wrapped with the Context wrappers, and everything works fine.

**Comment by robertbraeutigam** [comment]
Thanks for answering. Yes, "riddled" might be a bad way to put it. I wanted to say any invocation, even harmless sounding ones _may_ include threadlocals under the hood. Which is a problem, if you want to avoid them. Just now I'm working on a problem where a threadlocal was used in `startSpan`, because the parent was not set. Now spans appear under traces that have nothing to do with it.

Let me rephrase the problem. Seems to me (correct me if I'm wrong), that the SDK assumes, that a span will run on a single thread. This is probably the case for conventional Java applications, but reactive apps, CompletableFutures, cats-effect, etc. do not generally give you this guarantee.

In other words if you want to have a span for `foo().thenCompose(bar())`, how do you do it? You have no idea where `foo` is running, you have no idea where `bar` is running. They may be running on the same thread maybe not. Maybe running on the same thread pool, maybe not. What if `bar` wants to add some attributes, or start a child-span? Seems to me the only reasonable way to do this is to pass some object along. This should be the easy default. But it seems it's not. How would you do it?

**Reply by jkwatson** [reply]
We absolutely, in _no way_, assume that java applications will run on a single thread. This is why our Context implementation has many methods to assist with context propagation into other threads. 

If you are going to write manual instrumentation, you need to deeply understand how your application framework uses threads, and utilize the utilities we provide on the Context interface to make sure that the context is properly propagated. Our APIs will not do this for you, but they _will_ give you the tools you need, if you spend the time to understand them.

Context propagation is very tricky, and there is no "one size fits all" solution, which is why we provide you the tools, but do not try to solve every possible use-case.

If you want auto-instrumentation that will manage internal context propagation for you, please use the opentelemetry java-agent, which manages context propagation for most common frameworks.  If you want to do manual instrumentation, please study the Context interface and learn the utilities it provides, before assuming that we don't know what we're doing.

**Comment by jkwatson** [comment]
If your goal is completely avoid the use of ThreadLocals for context storage, you have two options: 
* Write a different ContextStorage implementation that fulfills your needs.
* Manually propagate your Context instances where you want them via method parameters, or some such. This is much like the way that all of golang works.

Both of these options will involve a _lot_ of work and will be very error prone. I recommend understanding how your framework spawns threads and use our Context utilities to make sure the spawning of them properly propagates the Context into the spawned threads.

**Reply by robertbraeutigam** [reply]
I choose option 2 for the project I'm involved in. I try to manually propagate `Context` instances. Seems to me, this is should be theoretically way easier, more explicit and less error prone, than trying to anticipate how the application will be mapped to threads (again, even a single span can be distributed among any number of threads, even multiple ones at a given time, then later joined to a single point where the span ends, and this is not controlled by any one framework).

This creates problems though, because it's not apparent which method invocations in this library will involve some threadlocals deep inside. I essentially have to go into the code and check that the execution path I'm on will not go to thread locals. This is obviously pretty error prone and time consuming. Also, I had to write a wrapper on top, because I can't give any of this directly to a developer, for the above reasons.

I understand your point, that this is not how this library was intended to be used. My point is though, that "manually" propagating the context, i.e. just forwarding the context as a parameter, is the widest possible abstraction, on which all the "magic" / automatic propagation strategies can be implemented on if needed. If that is right, it might be possible to separate the "essential" part of the library, which is a much smaller, much simpler, less involved codebase, and still provide convenience support for frameworks based on it on a higher layer.

This would have the benefits of:
- Supporting my use-case :), of not relying on automatic propagation, because we're using reactive, or fibers, or third-party libraries with their own stuff, or just want to make things explicit.
- Make it easier to differentiate / understand the "essential" parts of the library
- Potentially make the library more modular, i.e. layered

Obviously this is all just my experience based on the last days and weeks trying to make this work. It is intended to be feedback, not judgement.

**Reply by jkwatson** [reply]
I think if you just skip calling `Span.current()` or `Context.current()`, and always explicitly assign a parent Context to your created Spans, you will never hit a ThreadLocal. 

This isn't the default behavior, because most users want to let the agent handle the thread hops (or just accept the ThreadLocalStorage and make sure that your thread spawning is always wrapped with a Context propagating wrapper). You may be different, and that's ok, but you're not the primary use-case we have designed for.

Our mantra is to make the most common use-case simple, and the more esoteric possible, which we have.

**Reply by trask** [reply]
> you're not the primary use-case we have designed for

I don't necessarily think this is true. We did discuss reactive programming at length when designing the OpenTelemetry context.

@robertbraeutigam you'll see this reflected in all of the APIs accepting an explicit `Context`, and you can confirm this by checking that the OpenTelemetry SDK never calls `Context.makeCurrent()`, making thread locals entirely opt-in by users.

@robertbraeutigam it would help if you could post an example of where ThreadLocals are leaking in to your application, maybe they are coming from OpenTelemetry instrumentation as opposed to the OpenTelemetry SDK?

**Reply by jkwatson** [reply]
@trask if you don't explicitly override the parent, the SDK will ask for Span.current() as the parent when the span starts. So, this one is not necessarily obvious. And, I think my statement stands, that our _easiest_ API usage does assume that you're being using the ThreadLocal implementation of ContextStorage, and if you aren't careful, it'll be used to parent you spans implicitly.

We definitely make it _possible_ to not use the ThreadLocals, but it's not the easiest, most obvious path through the API.

**Reply by trask** [reply]
> @trask if you don't explicitly override the parent, the SDK will ask for Span.current() as the parent when the span starts.

Sorry, part of my point was that if you never call `makeCurrent`, then this usage has no effect since it's always the root (empty) context.

**Reply by johnpoth** [reply]
@robertbraeutigam thanks for this thread ! I am having exactly the same problem. Can't really add anything more than what you've said.

@jkwatson @trask could you show how to use the library without the ThreadLocals ?

Thanks !

**Reply by trask** [reply]
hi @johnpoth! all of the OpenTelemetry Java APIs allow you to pass in `Context` explicitly (so that you don't need to use ThreadLocals), e.g.

https://github.com/open-telemetry/opentelemetry-java/blob/ee7fd2744949ae0c060d28c73edfca56633dc170/api/all/src/main/java/io/opentelemetry/api/trace/SpanBuilder.java#L99-L110

**Comment by johnpoth** [comment]
@trask Thank you so much for your response!

I see that your snippet uses Context.current() which does call some ThreadLocalStorage. Will that pose some problems? In my case, some Spans will start and end in different threads.

**Reply by trask** [reply]
ah, sorry, you don't need to call `Context.current()`, just pass in the `Context` instance that you are passing around explicitly

**Comment by johnpoth** [comment]
@trask Right hence my confusion. The methods in _Context_ to create a new _Context_ Object seem to all refer to the ThreadLocal stuff. How to handle the Context lifecycle without the ThreadLocals ? Thank you so much !

**Reply by trask** [reply]
just replace `Context.current()` in the above code with `Context.root()`

**Reply by jkwatson** [reply]
Important to remember: all Context instances are completely immutable, and Context.root() is an empty Context instance, so you don't need to create a new one...just use the `root()` which is empty and immutable.

**Reply by johnpoth** [reply]
@trask @jkwatson thank you so much! So if my understanding is correct I should keep track of both the current Span and the current Context ? Or can I just use Context.root() everywhere and just keep track of the current Span like above
 
`
builder.setParent(Context.root().with(parent()))  
`

I would also search ContextPropagators when I don't find a parent Span...

Thanks !

**Reply by jkwatson** [reply]
Yes, I recommend only propagating a Context, and putting the current Span into the Context, using the normal methods.  
`context.with(span)`

**Reply by johnpoth** [reply]
Thanks @jkwatson ! I wasn't able to get the [log4j appender working](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-appender-1.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/log4j/appender/v1_2/LogEventMapper.java#L110) as it seems to use the Context.current() stuff. I there a way to give the current Context in the Log event ? Thanks !

**Reply by jkwatson** [reply]
The only way I know of might be to use the log4j MDC, but that also uses a ThreadLocal as the storage.

**Reply by johnpoth** [reply]
@jkwatson thank you so much for your response! I'll check that out. In the meantime I've let myself use makeCurrent() for that purpose when I start and end spans:

`
try (Scope ignored = context.makeCurrent()) {

    Span span = builder.setParent(context).startSpan();
    return context.with(span);
}
`

I've noticed however that the logEvent is triggered when startSpan() is called and so I get 0000 for traceId and SpanId in the logs on the first run. I figure startSpan() should be called after something like `context.with(span).makeCurrent()` but it's sort of a chicken and the egg problem as the Span is only available once startSpan() is called..... Thanks!

**Reply by johnpoth** [reply]
@jkwatson @trask so I've finally gotten rid of all the ThreadLocal stuff in the code base! However I've noticed [1] that the many libraries and frameworks supported in the OpenTelemetry ecosystem do rely on the OpenTelemetry Context being set in the ThreadLocal storage.

Is there a workaround for that ? I feel like the answer is no but you never know :)

Thanks !

[1] https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/bootstrap/Java8BytecodeBridge.java#L22

**Reply by jkwatson** [reply]
This would be a question for the instrumentation side of things. My guess is that it would take quite a bit of re-work of the instrumentation, but it's not my area of expertise, so I recommend asking over in the instrumentation repository.

**Reply by trask** [reply]
> Is there a workaround for that?

it depends, can you be more specific about the issue you are facing?
